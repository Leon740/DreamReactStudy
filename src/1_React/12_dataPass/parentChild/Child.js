import React from "react";
import PropTypes from "prop-types";

const Child = ({ carName }) => {
  return (
    <div>
      <p>
        My <u>Dream</u> is <b>{carName}</b>
      </p>
    </div>
  );
};

Child.propTypes = {
  carName: PropTypes.string.isRequired,
};

export default Child;
