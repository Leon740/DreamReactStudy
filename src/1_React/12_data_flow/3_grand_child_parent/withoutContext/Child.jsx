import React from 'react';
import PropTypes from 'prop-types';
import GrandChild from './GrandChild';

function Child({ carName, setCarName }) {
  return (
    <>
      <code>Child :</code>
      <p>
        My
        {' '}
        <u>Dream</u>
        {' '}
        is
        {' '}
        <b>{carName}</b>
      </p>

      <GrandChild carName={carName} setCarName={setCarName} />
    </>
  );
}

Child.propTypes = {
  carName: PropTypes.string.isRequired,
  setCarName: PropTypes.func.isRequired,
};

export default Child;
