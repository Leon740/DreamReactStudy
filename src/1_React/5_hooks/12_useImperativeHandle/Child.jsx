import React, { forwardRef, useRef, useImperativeHandle } from 'react';

function Child(_, ref) {
  const rfInput1 = useRef();
  const rfInput2 = useRef();
  const rfInput3 = useRef();

  function fnInput1Red() {
    rfInput1.current.style.background = 'red';
  }
  function fnInput2Green() {
    rfInput2.current.style.background = 'green';
  }
  function fnInput3Blue() {
    rfInput3.current.style.background = 'blue';
  }

  // === Concept
  // Escape using useImperativeHandle
  // ref = parent ref
  // () => {  } = callback function which overrides the parent ref
  // [] = dependecies

  useImperativeHandle(ref, () => ({
    fnInput1Red: () => fnInput1Red(),
    fnInput2Green: () => fnInput2Green(),
    fnInput3Blue: () => fnInput3Blue(),
  }), []);

  return (
    <div ref={ref}>
      <input ref={rfInput1} />
      <input ref={rfInput2} />
      <input ref={rfInput3} />
    </div>
  );
}

export default forwardRef(Child);
