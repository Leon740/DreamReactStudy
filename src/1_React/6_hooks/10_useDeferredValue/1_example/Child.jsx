/* eslint-disable react/no-array-index-key */
import React from 'react';

function Child({ input }) {
  const LIST_SIZE = 20000;
  const list = [];

  // === Theory : complex logic
  for (let i = 0; i <= LIST_SIZE; i += 1) {
    list.push(input);
  }

  return (
    <ul>{list.map((item, index) => <li key={index}>{item}</li>)}</ul>
  );
}
export default Child;
