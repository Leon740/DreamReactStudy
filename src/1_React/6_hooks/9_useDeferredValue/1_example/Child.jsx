/* eslint-disable react/no-array-index-key */
import React from 'react';

function Child({ input }) {
  const list = [];

  // === Theory : complex logic
  for (let i = 0; i <= 20000; i += 1) {
    list.push(input);
  }

  return (
    <ul>{list.map((item, index) => <li key={index}>{item}</li>)}</ul>
  );
}
export default Child;
