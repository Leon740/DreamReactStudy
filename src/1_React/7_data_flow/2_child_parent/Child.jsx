import React from 'react';
import PropTypes from 'prop-types';

function Child({ carName, setCarName }) {
  return (
    <input value={carName} onChange={(event) => setCarName(event.target.value)} />
  );
}

Child.propTypes = {
  carName: PropTypes.string.isRequired,
  setCarName: PropTypes.func.isRequired,
};

export default Child;
