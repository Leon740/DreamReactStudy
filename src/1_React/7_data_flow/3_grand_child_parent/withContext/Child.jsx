import React, { useContext } from 'react';
import ComponentsContext from './ComponentsContext';
import GrandChild from './GrandChild';

function Child() {
  const { carName } = useContext(ComponentsContext);

  return (
    <>
      <code>Child :</code>
      <p>
        My
        {' '}
        <u>Dream</u>
        {' '}
        is
        {' '}
        <b>{carName}</b>
      </p>

      <GrandChild />
    </>
  );
}

export default Child;
