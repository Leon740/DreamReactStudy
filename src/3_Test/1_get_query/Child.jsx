import React from 'react';

function Child({ list }) {
  if (!list?.length) return null;

  return (
    <ul>
      {list.map((listItem, index) => <li key={index}>{listItem}</li>)}
    </ul>
  );
}

export default Child;
