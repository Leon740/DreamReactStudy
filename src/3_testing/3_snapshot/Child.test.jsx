import { render, screen } from '@testing-library/react';
import Child from './Child';

// === Concept
// The idea of snapshot testing is to create a snapshot at the end of work (version of work)
// After updates compare old and new version of snapshots
// Possible options in this case: * update code to match old snapshot, * update old snapshot with new version of code

describe('List', () => {
  test('Snapshot', () => {
    const listData = ['React', 'Redux', 'TypeScript'];
    render(<Child list={listData} />);

    const list = screen.getByRole('list');

    expect(list).toBeInTheDocument();

    expect(list).toMatchSnapshot();
  });
});
