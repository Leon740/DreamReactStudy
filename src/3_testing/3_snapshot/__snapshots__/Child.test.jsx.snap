// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Snapshot 1`] = `
<ul>
  <li>
    React
  </li>
  <li>
    Redux
  </li>
  <li>
    TypeScript
  </li>
</ul>
`;
